cpp/src/arrow/python/util/datetime.h

```diff
@@ -69,6 +71,48 @@ static inline Status PyTime_from_int(int64_t val, const TimeUnit::type unit,
   return Status::OK();
 }
 
+static inline Status PyDateTime_from_int(int64_t val, const TimeUnit::type unit,
```

Comment:
This is really gnarly, we're going to have to root around in the Python datetimemodule implementation to get this right across platforms: https://github.com/python/cpython/blob/master/Modules/_datetimemodule.c#L4326. I started looking at this but it's a can of worms. We need to also avoid UNIX y2k problems